### PR TITLE
fix: deprecate recurringJobs from volume spec

### PR DESCRIPTION
**Pull request by weizhe0422**
[Longhorn 2865](https://github.com/longhorn/longhorn/issues/2865)

Signed-off-by: Ray Chang <ray.chang@suse.com>

 ### e2e test result
<details> 

   - test_basic 
     test_basic
  - test_recurring_job
     test_recurring_job
  - test_provisioner
     test_provisioner
   - test_csi
    test_csi
test_kubernetes py
  - test_orphan
    test_orphan py

</details>

### CODE REVIEW

**Comment by derekbit** [comment]
Is there a way to prevent users from upgrading Longhorn from v1.1 to v1.2?
Any issue after removing volume.Spec.RecurringJobs from the upgrade path if no?

**Comment by weizhe0422** [comment]
Hi @derekbit, I installed v1.1.0 with 5Gi **detached** volume (if not detached, the manage will be in crash loop for apply default settings.), and create snapshot/backup with 5Gi data.


Then upgrade to the versions without `volume.Spec.RecurringJobs`, and pass the e2e test results includes test_basic, test_recurringjobs, test_provisioner.

File: k8s/pkg/apis/longhorn/v1beta1/volume.go [hunk 1]
```diff
@@ -164,9 +150,6 @@ type VolumeSpec struct {
 
 	// Deprecated. Rename to BackingImage
 	BaseImage string `json:"baseImage"`
-
-	// Deprecated. Replaced by a separate resource named "RecurringJob"
-	RecurringJobs []VolumeRecurringJobSpec `json:"recurringJobs,omitempty"`
```
[the remark this answers is not in the document]
Comment (weizhe0422) [reply k8s/pkg/apis/longhorn/v1beta1/volume.go]: Understood. I have added `recurringJobs` back to v1beta1.

**Comment by shuo-wu** [comment]
Is there an official guideline/principle for the CRD field deprecation? Can we do it without CRD API version bumping?

**Comment by weizhe0422** [comment]
I checked the official documentation and found a section on how to deprecate fields of REST resources.
https://kubernetes.io/docs/reference/using-api/deprecation-policy/#fields-of-rest-resources

Although it is not specific to CRDs, I thought it might be somewhat relevant.

**Comment by shuo-wu** [comment]
It seems that the deprecated field can be either renamed in the new API version, or be entirely removing 

IIUC, for example, field `volume.spec.recurringJobs` in v1beta2 can be renamed as `volume.spec.deprecatedRecurringJob` (but I think we need to add a conversion webhook) rather than directly removed in v1beta3. And once Longhorn removes v1beta2 support, we can remove `volume.spec.deprecatedRecurringJob` in v1beta3. In other words, we cannot directly remove the deprecating field `volume.spec.recurringJobs` now, at least from the CRD definition, right?

**Comment by weizhe0422** [comment]
Following what you said, I tried to implement it in several steps. Wondering if this is different from what you think, please?

1. Rename `volume.spec.recurringJobs` in v1beta2 as `volume.spec.deprecatedRecurringJob`
2. In [`Volume` CRD conversion function](https://github.com/longhorn/longhorn-manager/blob/60df31ab85f5f561cc012685b0f40ae7642cb29d/k8s/pkg/apis/longhorn/v1beta1/volume.go#L231-L264), convert the `volume.spec.recurringJobs` in v1beta1 to `volume.spec.deprecatedRecurringJob` in v1beta2
    ```golang
    if err := copier.Copy(&vV1beta2.Spec.deprecatedRecurringJob, &v.Spec.RecurringJobs); err != nil {
	    return err
    }
    ```
3. If v1beta1 is retired from Longhorn in the future, `volume.spec.deprecatedRecurringJob` can also be removed in the next release.

**Comment by shuo-wu** [comment]
Not sure if we can do this kind of renaming in v1beta2. What do you think @innobead ?

**Pull request closed without merging**